LSPS0/README.md

```diff
@@ -0,0 +1,603 @@
+> compression greatly simplifies implementation, testing,
+> interoperability, and dependencies.
+
+The client:
+
+* MUST send single complete JSON-RPC 2.0 request objects,
+  UTF-8-encoded, as payloads for message ID 37913.
+* MUST NOT send JSON-RPC 2.0 notification objects
+  (i.e. every object it sends must have an `id` field).
+* MUST NOT use by-position parameter structures, and MUST
+  use by-name parameter structures.
+* MUST NOT batch multiple JSON-RPC 2.0 request objects in
+  an array as described in the "Batch" section of the
+  JSON-RPC 2.0 specification.
+
+> **Rationale** By disallowing by-position parameter
+> structures, other LSPS specifications need only to define
+> parameter names and not some canonical order of parameters
+> for by-position use.
+> Having to handle only by-name parameter structures also
+> simplifies the LSP code, as it does not have to check
+> whether the `params` value is an array or a dictionary,
+> and to separately map array elements to `params` keys.
+> Batched requests require batched responses according to
+> JSON-RPC 2.0, and it may be simpler for LSPs to handle
+> unrelated LSPS methods separately without requiring
+> re-batching of the responses; it gives a simple "one
+> message is one request / response" rule.
+
+The LSP:
+
+* MUST send single complete JSON-RPC 2.0 response objects,
+  UTF-8-encoded, as payloads for message ID 37913.
+* MAY send JSON-RPC 2.0 notification objects
+  (i.e. it may send an object without `id` but with `method`,
+  `params`, and `jsonrpc` fields).
+  * MUST use by-name parameter structures for notifications.
+* MUST respect the [JSON-RPC 2.0 standard error codes][].
+* SHOULD NOT send message ID 37913 unless the peer already sent a
+  message ID 37913, possibly in a past connection session.
+
+[JSON-RPC 2.0 standard error codes]: https://www.jsonrpc.org/specification#error_object
+
+> **Rationale** The peer sending message ID 37913 is an
+> indicator that it understands the LSPS0 protocol and wishes
+> to act as a client.
+> If the peer does not send it, then it is not an LSPS client
+> and the LSP has no reason to send message ID 37913.
+> Notifications allow the LSP to initiate with the client,
+> such as if the client has an incoming payment that the LSP
+> learns about.
+
+JSON-RPC 2.0 is an asynchronous protocol, and the order
+in which the LSP sends responses MAY be different from the
+order in which the client sent the requests.
+
+Other LSPS specifications:
+
+* MUST indicate `method` names for each API endpoint and each
+  LSP-initiated notification, as well as the key names of the
+  `params` for each `method`, and the meanings of each
+  parameter.
+  * MUST prefix `method` names:
+    * For `method`s for client requests, prefix with `lsps`
+      followed by the LSPS number followed by `.`, e.g.
+      `lsps999.dothis`.
+    * For `method`s for LSP notifications, prefix with
+      `n.lsps` followed by the LSPS number followed by `.`,
```

Comment:
We can imagine that there might some day be a client request `method` whose name starts with `n.`, so that would cause conflict between the `lsps999.${METHOD}` and `lsps999.n.${NOTIFICATION}`.  Unlikely but this separates the namespaces better.

Might as well separate them, as requests are always from client -> LSP while notifications are always from LSP -> client, so you could check that everything from a client has `method` matching the regex `^lsps[0-9]+\.` and every notification from an LSP has `method` matching the regex `^n\.lsps[0-9]+\.`

Reply:
Point is not message parsing, the point is internal checks in your code that you have not misspelled `lsps0.listprotocols` as `lssp0.listprotocols` or some such.  Separating notifications supports this so you do not accidentally use a notification name for an API name in your code, especially if you do stuff like make constant variables that contain the method name.  This is more a quick-and-dirty code-side check than something you do on parsing.

Reply:
I tend to agree with @tnull in this case. Each LSPSx spec needs to make sure it doesn't have conflicting names anyway. So the `n.` is unnecessary.

Reply:
Done.